File: 链表/移除链表元素/203.js
/*
 * @Description: 
 * @Author: young
 * @Date: 2022-08-22 20:26:38
 * @LastEditTime: 2022-08-22 20:31:40
 * @FilePath: \young_leetcode\链表\移除链表元素\203.js
 */

// 链表节点的构造函数
function ListNode(val, next) {
    this.val = (val === undefined ? 0:val);
    this.next = (next === undefined ? null:next);
}

const removeElements = function(head, val) {
    const head_node = new ListNode(0, head);
    let pointer = head_node;

    while (pointer.next) {
        if (pointer.next.val === val) {
            pointer.next = pointer.next.next;
            continue;
        }
        pointer = pointer.next;
        
    } 

    return head_node.next;
}

console.log(removeElements([1,2,6,3,4,5,6],6))

File: 链表/移除链表元素/203.c
/*
 * @Description: 
 * @Author: young
 * @Date: 2022-08-22 20:18:29
 * @LastEditTime: 2022-08-22 20:25:39
 * @FilePath: \young_leetcode\链表\移除链表元素\203.c
 */

struct ListNode {
    int val;
    struct ListNode *next;
};


struct ListNode *removeElements(struct ListNode* head, int val) {
    struct ListNode *temp;

    while (head && head->val == val) {
        temp = head;
        head = head->next;
        free(temp);
    }

    struct ListNode *pointer = head;

    while (pointer && (temp = pointer->next)) {
        if (temp->val == val) {
            pointer -> next = temp -> next;
            free(temp);
        }
        else {
            pointer = pointer -> next;
        }
    }

    return head;
}

